spec/prog/postgres/converge_postgres_resource_spec.rb

```diff
@@ -3,403 +3,484 @@
-      parent = instance_double(PostgresResource, ready_for_read_replica?: false)
-      allow(postgres_resource).to receive_messages(read_replica?: true, parent: parent)
+      parent = PostgresResource.create(
+        name: "pg-parent", superuser_password: "dummy-password", ha_type: "none",
+        target_version: "17", location_id: location_id, project_id: project.id,
+        user_config: {}, pgbouncer_user_config: {}, target_vm_size: "standard-2",
+        target_storage_size_gib: 64, private_subnet_id: private_subnet.id
+      )
+      Strand.create_with_id(parent, prog: "Postgres::PostgresResourceNexus", label: "wait")
+      postgres_resource.update(parent_id: parent.id)
+      allow(parent).to receive(:ready_for_read_replica?).and_return(false)
+      allow(postgres_resource).to receive(:parent).and_return(parent)
 
       expect { nx.start }.to nap(60)
     end
 
     it "registers a deadline and hops to provision_servers if read replica parent is ready" do
-      parent = instance_double(PostgresResource, ready_for_read_replica?: true)
-      expect(nx).to receive(:register_deadline).with("wait_for_maintenance_window", 2 * 60 * 60)
-      allow(postgres_resource).to receive_messages(read_replica?: true, parent: parent)
+      parent = PostgresResource.create(
+        name: "pg-parent2", superuser_password: "dummy-password", ha_type: "none",
+        target_version: "17", location_id: location_id, project_id: project.id,
+        user_config: {}, pgbouncer_user_config: {}, target_vm_size: "standard-2",
+        target_storage_size_gib: 64, private_subnet_id: private_subnet.id
+      )
+      Strand.create_with_id(parent, prog: "Postgres::PostgresResourceNexus", label: "wait")
+      postgres_resource.update(parent_id: parent.id)
+      allow(parent).to receive(:ready_for_read_replica?).and_return(true)
+      allow(postgres_resource).to receive(:parent).and_return(parent)
 
+      expect(nx).to receive(:register_deadline).with("wait_for_maintenance_window", 2 * 60 * 60)
       expect { nx.start }.to hop("provision_servers")
     end
   end
 
   describe "#provision_servers" do
     before do
-      allow(postgres_resource).to receive(:has_enough_fresh_servers?).and_return(false)
-      allow(postgres_resource.servers[0]).to receive(:vm).and_return(instance_double(Vm, vm_host: instance_double(VmHost, id: "vmh-id-1")))
-      allow(postgres_resource.servers[1]).to receive(:vm).and_return(instance_double(Vm, vm_host: instance_double(VmHost, id: "vmh-id-2")))
+      strand.update(label: "provision_servers")
+      create_server(representative: true, vm_host_data_center: "dc1")
+      create_server(representative: false, vm_host_data_center: "dc2")
     end
 
     it "hops to wait_servers_to_be_ready if there are enough fresh servers" do
-      expect(postgres_resource).to receive(:has_enough_fresh_servers?).and_return(true)
+      allow(postgres_resource).to receive(:has_enough_fresh_servers?).and_return(true)
```

Comment:
This seems wrong. `allow` can make sense in `before` hooks, where it may or may not be called. For individual specs, it should probably be avoided, since you should know in the spec whether the method would be called.  Similar issue in other specs.

I'm not sure whether the mocking here is actually necessary. The related methods seem to only do database access.